tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc

```diff
@@ -2797,5 +2980,563 @@ llvm::Optional<Value> convertTFConv2DCommon(
+      rewriter, op, flattened_coeff_vec);
+
+  // Multiply the coefficients by the coordinates
+  auto flattened_indices_mul_op = rewriter.create<tosa::MulOp>(
+      op->getLoc(),
+      RankedTensorType::get(ArrayRef<int64_t>(indices_matrix_shape),
+                            indices_type.getElementType()),
+      indices_matrix_reshape_op.getResult(), flattened_coeff_value, 0);
+
+  // Sum up the products of the coefficients and coordinates
+  auto flattened_indices_reduce_op = rewriter.create<tosa::ReduceSumOp>(
+      op->getLoc(),
+      RankedTensorType::get(ArrayRef<int64_t>(tosa_indices_shape),
+                            indices_type.getElementType()),
+      flattened_indices_mul_op.getResult(), rewriter.getI64IntegerAttr(1));
+
+  // And reshape to [N, W]
+  auto tosa_indices_reshape_op = rewriter.create<tosa::ReshapeOp>(
+      op->getLoc(),
+      RankedTensorType::get(ArrayRef<int64_t>(tosa_indices_shape),
+                            indices_type.getElementType()),
+      flattened_indices_reduce_op.getResult(),
+      rewriter.getI64ArrayAttr(tosa_indices_shape));
+
+  // Now the gather op itself
+  auto tosa_gather_op = rewriter.create<tosa::GatherOp>(
+      op->getLoc(),
+      RankedTensorType::get(ArrayRef<int64_t>(tosa_gather_result_shape),
+                            result_type.getElementType()),
+      tosa_values_reshape_op.getResult(), tosa_indices_reshape_op.getResult());
+
+  // Finally, reshape back to the original output shape of [Indices,
+  // ParamChannels].
+  return rewriter
+      .create<tosa::ReshapeOp>(op->getLoc(), result_type,
+                               tosa_gather_op.getResult(),
+                               rewriter.getI64ArrayAttr(result_type.getShape()))
+      .getResult();
+}
+
+// Lowers OneHot operator to a sequence of TOSA ops.
+llvm::Optional<Value> convertOneHotOp(PatternRewriter& rewriter, Operation* op,
+                                      Value result_value, Value indices_value,
+                                      Value on_value, Value off_value,
+                                      int32_t depth, int32_t axis) {
+  auto result_type = result_value.getType().dyn_cast<RankedTensorType>();
+  auto indices_type = indices_value.getType().dyn_cast<RankedTensorType>();
+  auto on_value_type = on_value.getType().dyn_cast<RankedTensorType>();
+  auto off_value_type = off_value.getType().dyn_cast<RankedTensorType>();
+
+  if (!result_type) return llvm::None;
+  if (!indices_type) return llvm::None;
+  if (!on_value_type) return llvm::None;
+  if (!off_value_type) return llvm::None;
+
+  // OneHot operator creates a new tensor with shape indices.shape[:axis] +
+  // [depth] + indices.shape[axis:] For each index in 'indices', it needs to be
+  // within range of [0, depth - 1] and the [..., k, ...] = on_value (if k =
+  // index), or [..., k, ...] = off_value (if k != index)
+  //
+  // The lowering below assumes depth is always known at compile time.
```

Comment:
Where are you checking this condition and early exiting? (forgive me if I missed it)

Reply:
No I didn't. I will add the check in next round.

Reply:
Oh actually we did. In legalize_tf.cc and legalize_tfl.cc, we do return failure() if matchPattern() fails.